courses/management/commands/generate_certificate.py

```diff
@@ -0,0 +1,65 @@
+"""
+Management command to generate certificates
+"""
+from django.core.management.base import BaseCommand, CommandError
+from courses.utils import sync_courserun_grades, sync_courserun_certificates
+
+
+class Command(BaseCommand):
+    """
+    Command to generate certificates for a course.
+    """
+
+    help = "Generates certificates for a course for all enrolled users or a specified user."
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            "--email",
+            type=str,
+            help="The email address of the enrolled User.",
+            required=False,
+        )
+        parser.add_argument(
+            "--course-run",
+            type=str,
+            help="The 'courseware_id' value for a CourseRun.",
+            required=True,
+        )
+        parser.add_argument(
+            "--grade",
+            type=float,
+            help="Override a grade. Setting grade: 0 blacklists a user. Setting a passing grade \
+                (>0.0) whitelists a user. range: 0.0 - 1.0",
+            required=False,
+        )
+        parser.add_argument(
+            "--force",
+            action="store_true",
+            help="Force overwrite of local grade record.",
+        )
+        super().add_arguments(parser)
+
+    def handle(self, *args, **options):
```

Comment:
I had some major changes in mind for this handler function, so it was easier to work through on it in my IDE and paste the results here. I didn't test this – I just wanted to make sure the changes looked right to me in general before making a suggestion. There are some references to suggestions I made in the overall PR comment. Here it is:

```python
    def handle(self, *args, **options):
        """Handle command execution"""
        # Validate input and configuration
        if options["grade"] and not options["user"]:
            raise CommandError(
                "No user supplied with override grade. Force overwrite of grade is not supported "
                "for all users. Grade should only be supplied when a specific user is targeted."
            )
        user = fetch_user(options["user"]) if options["user"] else None
        try:
            run = CourseRun.objects.get(courseware_id=options["run"])
        except CourseRun.DoesNotExist:
            raise CommandError(
                "Could not find run with courseware_id={}".format(
                    options["run"]
                )
            )
        if options["grade"]:
            new_grade = float(options["grade"])
            if new_grade and (new_grade < 0.0 or new_grade > 1.0):
                raise CommandError("Invalid value for grade. Allowed range: 0.0 - 1.0")
            should_blacklist = new_grade == 0.0
        else:
            new_grade = None
            should_blacklist = False
        should_update = options["update"]
        # !!! PR NOTE: You can redefine this and/or supply arguments as necessary
        _verify_configuration()

        # Fetch user(s) and edX grade data
        edx_client = get_edx_api_grades_client()
        if user:
            edx_grade = edx_client.current_grades.get_student_current_grade(
                user.username, run.courseware_id
            )
            user_edx_grade_iter = zip([edx_grade], [user])
        else:
            self.stdout.write(
                "No user specified. Attempting to sync grades and generate certificates for all applicable users"
            )
            user_edx_grade_iter = get_edx_grades_with_users(edx_client, run)

        # Loop users and edX grades and make necessary changes to local grade records and certificates
        results = []
        for current_edx_grade, user in user_edx_grade_iter:
            result = []
            course_run_grade, created, updated = self.ensure_course_run_grade(
                user=user,
                course_run=run,
                edx_grade=current_edx_grade,
                should_update=should_update,
            )
            if created:
                result.append("New grade record created")
            elif updated:
                result.append("Grade record updated with new data from edX grade API")
            if new_grade:
                course_run_grade.grade = new_grade
                course_run_grade.passed = bool(new_grade)
                course_run_grade.letter_grade = None
                course_run_grade.save_and_log(None)
                result.append("Grade overriden - new grade value: {}".format(course_run_grade.grade))

            if should_blacklist:
                deleted_count = CourseRunCertificate.objects.filter(
                    user=course_run_grade.user, course_run=course_run_grade.course_run
                ).delete()
                if deleted_count:
                    result.append("Certificate deleted")
            else:
                certificate, created = CourseRunCertificate.objects.get_or_create(
                    user=course_run_grade.user, course_run=course_run_grade.course_run
                )
                result.append(
                    "Certificate created"
                    if created else
                    "Certificate ignored (already existed)"
                )
            result_summary = ", ".join(result) if result else "No changes made"
            results.append(
                "Processed user {} ({}) in course run {}. Result: {}".format(
                    user.username,
                    user.email,
                    run.courseware_id,
                    result_summary
                )
            )

        # Report results
        for result in results:
            self.stdout.write(
                self.style.SUCCESS(result)
            )
```

Reply:
Updated. Some points below:

- I created two methods for the grades api: `get_edx_grades_api_client` and `get_edx_grades_with_users` as suggested by you. Both are in `courseware.api` and are used to fetch grades from OpenEdX.
- In `courseware.utils` I have created `ensure_course_run_grade`, `ensure_course_run_grade_certificate` and `ensure_course_run_certificate`. I have not attached them specifically to the command class because this needs to be done in another ticket (#877) so I want to avoid code duplication. @asadiqbal08 can reuse these methods in his PR as well.